### PR TITLE
[bugfix] Fixes #17576 and #17574 Add Z support and fixes crashes

### PR DESCRIPTION
**Pull request by pblottiere**
## Description

This PR fixes https://issues.qgis.org/issues/17574 and https://issues.qgis.org/issues/17576 by adding Z support to some map tools and fixing some crashes.

## Checklist

- [x] Commit messages are descriptive and explain the rationale for changes
- [x] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [ ] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [ ] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and containt sufficient information in the commit message to be documented
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTE.md#contributing-to-qgis) before each commit

### CODE REVIEW

**Comment by pblottiere** [comment]
Hi @lbartoletti!

Can you give me your opinion on this PR if you have some time please? Just to ensure that I didn't mess up the original behavior of these map tools...

Thank you :)

**Comment by lbartoletti** [comment]
Hi @pblottiere!

That sounds good to me!

I'll look into it in depth soon. I'm thinking of giving you a full answer by middle of the next week.

Overall on QGis, I think there is still a lot to say about Z and 3D for these (and other) tools. Some methods have a 3D support and others not.
What to do in these cases? Use, the average coordinate, find the reference plane, take the first 3D point (seems your choice?), if all do not have 3D stay in 2D, separate 2D from 3D methods (and what about CSG / polyhedron 😉 ), etc.?
FYI, I am currently testing some methods ([3d/skew lines intersection](https://github.com/lbartoletti/qgis/tree/3d_line_intersection)) that will improve 3D support.

In addition, the information tools also have no 3D support. Test with this WKT (LINESTRINGZ(30 10 5, 10 30 5, 40 40 10) ) Identify results tool, measure tool and field calculator ($length) return a length of 59.91 instead of 60.30.

See you soon.

**Comment by pblottiere** [comment]
> Overall on QGis, I think there is still a lot to say about Z and 3D for these (and other) tools. Some methods have a 3D support and others not.
What to do in these cases? Use, the average coordinate, find the reference plane, take the first 3D point (seems your choice?), if all do not have 3D stay in 2D, separate 2D from 3D methods (and what about CSG / polyhedron :wink: ), etc.?

I'm glad you asked because I wanted to address this point :).

For now, I took the first 3D point for getting the z coordinate of:
- the center of a QgsCircle from 3 points
- a line intersection
- the center of a QgsTriangle

I just made an average on Z coordinates in case of the Segment Mid Point method.

But using an average for each method would be a first step. Of course, I am open to suggestions :).

**Comment by lbartoletti** [comment]
Hi @pblottiere!

So, classes are mostly built for 2D - i. e. without having different Z's on the shape / no inclination member.
Most methods can be on a horizontal plane with a certain Z (often coming from the center or altitude of a middle point).
However, the lengths are always calculated on the 2D coordinates (same horizontal plane).

It is possible to propose methods to improve 3D support.
I designed the methods in anticipation of possible 3D uses (inclination) which means that for 2D there can be differences between the methods (but they are documented).
For the GUI part, I stayed on a basic 2D usage letting the methods and drawing tools inherited from QGis.
Here is a small table summarizing this, with the AutoCAD comparison, what can be proposed in these drawing tools and some thought for 3D (Is it planned to draw in the 3D view?).


Method | Z from class | AutoCAD | 2D GUI | Thought/is it possible in 3D (with inclination)? |
---------|-------------|-----------|---------|------------------------------------------------
QgsEllipse (center, semiMajorAxis, semiMinorAxis, azimuth) | Z from center | Z from first point (center) | Z from center | - (Z from center and a plane?) |
QgsEllipse::fromCenter2Points | Z from center | Z from first point (center) | Z from center | Z from center Plane defined by the center and other points? |
QgsEllipse::fromCenterPoint | Z from center | - | Z from center / Z from the first 3D point | - (Z from center and a plane?) |
QgsEllipse::fromExtent | midpoint (average) | - | Z from first point / Z from the first 3D point | - (Z from midpoint and a plane ?) |
QgsEllipse::fromFoci | midpoint (average) | - | Z from first point / Z from the first 3D point | Z from midpoint foci. Plane defined by foci and the point? |
QgsCircle (fromCenterRadius) | Z from center | Z from first point (center) | Z from center | - (Z from center and a plane?) |
QgsCircle::from2Points | midpoint (average) | Z from first point | Z from first point / Z from the first 3D point | - (Z from midpoint and a plane?) |
QgsCircle::from3Points | drop Z value | Z from first point | Z from first point / Z from the first 3D point / Z from plane defined by 3 points | Z from plane defined by 3 points |
QgsCircle::from3Tangents | drop Z value | Only if start and end points on segments have the same Z (can have different Z on segments) ; drop Z value | drop Z value | ? |
QgsCircle::from2TangentsPoint | drop Z value | Only if start and end point of segments have the same Z ; drop Z value | drop Z value | ? |
QgsCircle::fromCenterDiameter | Z from center | Z from first point (center) | Z from center | - (Z from center and a plane?) |
QgsCircle::fromCenterPoint | Z from center | Z from first point (center) | Z from center / Z from the first 3D point | - (Z from center and a plane?) |
QgsCircle::fromExtent | midpoint (average) | - |  Z from first point / Z from the first 3D point | - (Z from midpoint and a plane?) |
QgsCircle::minimalCircleFrom3Points | drop Z value | - | drop Z value (not present) | See minimalSphere? |
QgsRegularPolygon (fromCenter...) | Z from center | Z from first point (center) | Z from first point / Z from the first 3D point | - (Z from center and a plane?) |
QgsRegularPolygon (2 points...) | Z from center (projection from first point) | Z from first point | Z from first point / Z from the first 3D point | - (Z from center and a plane?) |
Rectangle3Points | - | - | Like a linestring/polygon | Define a plane |
RectangleCenter | - | - | Z from first point / Z from the first 3D point | - (Z from center and a plane?) |
RectangleExtent | - | Z from center | Z from first point / Z from the first 3D point | - (Z from midpoint and a plane?) |

The average Z coordinate of the points (for 2D) can be a good solution if all points have a Z, this is not always the case. So it remains your proposal and the habit on AutoCAD.
AutoCAD considers that you determine the Z from the first point; if it doesn't, the Z is 0. Your solution, consider that we leave a chance to get a Z in the points. I don't have a philosophy on the subject, the second one is perhaps more consistent with the intention to use a Z.
However, we must be consistent with the tools proposed. 
I don't know if this is the right method, but in order to improve the classes, I will modify the mapTools instead.

I think that all drawing tools should behave in the same way, with some exceptions.
Taking up again your idea of taking the first 3D point, here's what I suggest:

* Leave it like this
QgsCircle() / fromCenterRadius
QgsCircle::fromCenterDiameter
QgsEllipse() / center and axis
Rectangle3Points

* Take the first 3D point and propagate its z
QgsCircle::from2Points
QgsCircle::from3Points
QgsCircle::fromCenterPoint
QgsCircle::fromExtent
QgsEllipse::fromCenter2Points
QgsEllipse::fromCenterPoint
QgsEllipse::fromExtent
QgsEllipse::fromFoci
RectangleCenter
RectangleExtent
CircularArcString methods

* Drop the Z, it seems to me difficult to take a Z in this case (a tangential point having a Z? it doesn't seem clear to me for the user)
QgsCircle::from2TangentsPoint
QgsCircle::from3Tangents

In any case, these tools should be well documented 😃 

Regards

**Comment by lbartoletti** [comment]
I have just tested by drawing lines (into memory layer and geopackage) by snaping on 3D points (point layer in a geopackage and postgis base) I always have a Z at 0.
Maybe I didn't do the right thing?

**Comment by pblottiere** [comment]
Hi @lbartoletti !

Firstly, thank you for your input :+1:! Secondly, sorry for the delay...

I think I took into account all your comments about taking the 1st point with a Z value, excepted about tangential points:

> Drop the Z, it seems to me difficult to take a Z in this case (a tangential point having a Z? it doesn't seem clear to me for the user)
> QgsCircle::from2TangentsPoint
> QgsCircle::from3Tangents

@haubourg what do you think about it? I don't have a strong opinion, it's just to have a second thought :).

Moreover, I added some unit tests for ellipse, circle, rectangle, circular string and regular polygon map tools about Z values.

**Comment by pblottiere** [comment]
> I have just tested by drawing lines (into memory layer and geopackage) by snaping on 3D points (point layer in a geopackage and postgis base) I always have a Z at 0.
> Maybe I didn't do the right thing?

@lbartoletti  What map tool did you use in this case?

File: src/core/geometry/qgscircle.cpp [hunk 1]
```diff
@@ -40,6 +40,8 @@ QgsCircle QgsCircle::from2Points( const QgsPoint &pt1, const QgsPoint &pt2 )
   double azimuth = QgsGeometryUtils::lineAngle( pt1.x(), pt1.y(), pt2.x(), pt2.y() ) * 180.0 / M_PI;
   double radius = pt1.distance( pt2 ) / 2.0;
 
+  QgsGeometryUtils::setZValueFromPoints( QgsPointSequence() << pt1 << pt2, center );
```
Comment (lbartoletti) [inline src/core/geometry/qgscircle.cpp]: Can you please update the documentation for all methods?

**Comment by lbartoletti** [comment]
Hi @pblottiere !

>>     I have just tested by drawing lines (into memory layer and geopackage) by snaping on 3D points (point layer in a geopackage and postgis base) I always have a Z at 0.
>>     Maybe I didn't do the right thing?
>> 
> @lbartoletti What map tool did you use in this case?

All tools.
Maybe, I misunderstood. QgsGeometryUtils::setZValueFromPoints takes the first Z, that seems fine, but QgsMapToolCapture::fromPointXY applies the default Z value from the settings, not from the snapped points?

Example:
Z default value: 1234

Two points:
p1: PointZ(1978952.0, 5197817.5)
p2: PointZ(1979079.9, 5197801.0 1004)
QgsCircle.from2Points(p1, p2) -> Z center = 1004

But with map tool Z center = 1234

When I draw a simple line, Z are taken from the snapped points.

File: src/core/geometry/qgscircle.cpp (hunk 1 above)
Comment (pblottiere) [reply src/core/geometry/qgscircle.cpp]: Good point. I'm doing to do that.

**Comment by pblottiere** [comment]
> Drop the Z, it seems to me difficult to take a Z in this case (a tangential point having a Z? it doesn't seem clear to me for the user)

@lbartoletti After talking with Régis, it seems pertinent to drop z dimension in case of tangential points as you proposed. I'm going to do that so.

**Comment by pblottiere** [comment]
>  I have just tested by drawing lines (into memory layer and geopackage) by snaping on 3D points (point layer in a geopackage and postgis base) I always have a Z at 0.
> ...
> Maybe, I misunderstood. QgsGeometryUtils::setZValueFromPoints takes the first Z, that seems fine, but QgsMapToolCapture::fromPointXY applies the default Z value from the settings, not from the snapped points?

> When I draw a simple line, Z are taken from the snapped points.

@lbartoletti Let's take an example to understand one another :).

1. create a new scratch layer with LineString geometry type and Z dimension
2. update default Z value to 111.0
3. add a new line
4. update default Z value to 222.0
5. add a second line thanks to some snapped points from the first line

The result for the first line:

z4

The result for the second line:

z2

In this case:
1. new points based on snapped points have z=111.0.  In this case, Z are taken from the snapped points.
2. new points **not** based on snapped points have z=222.0. In this case, Z are taken from settings (default value)

Do you agree with such a behavior? @haubourg Do you have an opinion?

**Comment by lbartoletti** [comment]
@pblottiere yes! It's works for the linestring tool, it will also work for these tools?

**Comment by lbartoletti** [comment]
🍾 Well done Paul :) ! 🍾

**Comment by pblottiere** [comment]
Thank you for your tests @lbartoletti!

I'll merge this PR tomorrow if I have no more comments in the meantime.

**Pull request merged**